### PR TITLE
DEP: special: deprecation warning for `sph_harm` + comments

### PR DESCRIPTION
**Pull request by steppi**
#### Reference issue

This is a promised follow-up to #22156 

#### What does this implement/fix?

This PR adds a deprecation warning for `sph_harm`. It is done by calling [PyErr_WarnEx](https://docs.python.org/3/c-api/exceptions.html#c.PyErr_WarnEx) from within the wrapper for `sph_harm` inside of `scipy/special/xsf_special.h`. This header is intended for SciPy specific modifications of `xsf` code, and as it turns, `sph_harm` already had a wrapper there in order to have a warning printed when floating point values are passed for arguments which should be integers.

I've also suppressed the warnings in the tests and added a test to check that the deprecation warning appears.

In addition, I've also added comments explaining why we specifically left in tests for the deprecated functions instead of updating them to use the new function, as promised here: https://github.com/scipy/scipy/pull/22156#discussion_r1895081141.

cc @h-vetinari, @jorenham 

#### Additional information

I'm fine if this gets squash merged.

### CODE REVIEW

**Comment by steppi** [comment]
Only failure is unrelated from the flaky fail slow tests:

```
(187043 durations < 0.25s hidden.  Use -vv to show these durations.)
=========================== short test summary info ===========================
FAILED scipy\stats\tests\test_stats.py::TestPearsonr::test_bootstrap_ci[less-numpy]
= 1 failed, 58888 passed, 3773 skipped, 191 xfailed, 14 xpassed in 469.30s (0:07:49) =
Error: Process completed with exit code 1.
```

**Comment by rgommers** [comment]
It's a bit late for this to be backported - normally the cutoff for that is before release branch creation and before rc1 (part of the point of release candidates is for downstream to be able to give feedback on bc-breaking changes). Is there a reason this must still go into 1.15 and not 1.16?

**Comment by h-vetinari** [comment]
> Is there a reason this must still go into 1.15 and not 1.16?

These methods were replaced and deprecated in 1.15, but only in documentation (due to a missed follow-up). I support a backport, unless it causes extraordinary effort for release mgmt. (which I don't expect).

**Comment by steppi** [comment]
I think I get @rgommers point though. Even though the intention was to deprecate `lpmn`, `clpmn`, `lpn`, and `sph_harm`, and the deprecation notices appear in the documentation, downstream maintainers may be annoyed if they've already tested rc1, only to find new warnings pop up in rc2. If that's the case though, then https://github.com/scipy/scipy/pull/22156 shouldn't be backported either. If the deprecation warnings don't go in until 1.16, then we could push back the removal time to 1.18.

**Comment by jorenham** [comment]
For what it's worth; `1.15.0rc1` has been downloaded approx `93_600` times, whereas `1.14.1` has been downloaded approx `85_140_000`:
https://pepy.tech/projects/scipy?timeRange=threeMonths&category=version&includeCIDownloads=true&granularity=daily&viewType=chart&versions=1.14.1.6%2C1.14.1.5%2C1.14.1.4%2C1.15.0rc1%2C1.14.1. Compared to other RC releases, that's an order of magnitude higher.

File: scipy/special/tests/test_data.py [hunk 1]
```diff
@@ -56,6 +56,15 @@ def data_local(func, dataname, *a, **kw):
     return FuncData(func, DATASETS_LOCAL[dataname], *a, **kw)
 
 
+# The functions lpn, lpmn, clpmn, and sph_harm appearing below are
+# deprecated in favor of legendre_p_all, assoc_legendre_p_all,
+# assoc_legendre_p_all (assoc_legendre_p_all covers lpmn and clpmn),
```
Comment (tylerjereddy) [inline scipy/special/tests/test_data.py:61]: `assoc_legendre_p_all` gets repeated doesn't it?

File: scipy/special/tests/test_legendre.py [hunk 2]
```diff
@@ -10,6 +10,14 @@
 from scipy.special import (legendre_p, legendre_p_all, assoc_legendre_p,
     assoc_legendre_p_all, sph_legendre_p, sph_legendre_p_all)
 
+# The functions lpn, lpmn, clpmn, appearing below are
+# deprecated in favor of legendre_p_all, assoc_legendre_p_all, and
+# assoc_legendre_p_all (assoc_legendre_p_all covers lpmn and clpmn)
```
Comment (tylerjereddy) [inline scipy/special/tests/test_legendre.py:15]: `assoc_legendre_p_all` repetition?

File: scipy/special/tests/test_mpmath.py [hunk 3]
```diff
@@ -671,6 +671,14 @@ def test_lambertw_smallz():
 # Systematic tests
 # ------------------------------------------------------------------------------
 
+# The functions lpn, lpmn, clpmn, and sph_harm appearing below are
+# deprecated in favor of legendre_p_all, assoc_legendre_p_all,
+# assoc_legendre_p_all (assoc_legendre_p_all covers lpmn and clpmn),
```
Comment (tylerjereddy) [inline scipy/special/tests/test_mpmath.py:676]: same here?

**Review by tylerjereddy — COMMENTED** [review]
I guess we can try to backport though not ideal as noted.

I added minor comments, but apart from that we should probably squash merge to make it easier for me to backport.

**Comment by rgommers** [comment]
> Compared to other RC releases, that's an order of magnitude more.

That's pretty remarkable indeed:

image

Maybe some package added a dependency on it, or perhaps people are really keen on the free-threading wheels:)

**Comment by jorenham** [comment]
> Maybe some package added a dependency on it, or perhaps people are really keen on the free-threading wheels:)

The dev version of `scipy-stubs` does this, but I somehow doubt that its CI could be responsible for tens of thousands of downloads 🤷🏻

**Comment by h-vetinari** [comment]
> or perhaps people are really keen on the free-threading wheels

I think that's the most plausible hypothesis TBH

File: scipy/special/tests/test_data.py (hunk 1 above)
Comment (steppi) [reply scipy/special/tests/test_data.py:61]: I guess it's not the clearest way to say `lpmn` and `clpmn` are both covered by `assoc_legendre_p_all`, and the parenthetical covers that already, so yeah, I'll change it.

Comment (tylerjereddy) [reply scipy/special/tests/test_data.py:61]: oh, I see what you did... ok, no this is fine then!

**Comment by tylerjereddy** [comment]
Ok, the diff looked "ok," release manager discretion to get this in a bit faster for RC2.

**Pull request merged**